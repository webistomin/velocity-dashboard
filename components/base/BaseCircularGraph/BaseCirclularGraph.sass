+b('base-circular-graph')
  position: relative
  width: 223px

  +e('base')
    display: block

  +e('fill')
    overflow: hidden
    position: absolute
    top: 0
    left: 0
    width: 0
    animation: progress-bar 2s ease-in-out

  +e('info')
    text-align: center
    position: absolute
    left: 50%
    top: 50%
    transform: translateX(-50%)

  +e('title')
    margin-bottom: rem(4px)
