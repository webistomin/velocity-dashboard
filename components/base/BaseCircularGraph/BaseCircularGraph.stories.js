import { number } from '@storybook/addon-knobs';
import BaseCircularGraph from './BaseCircularGraph';

export default {
  title: 'BaseCircularGraph',
  parameters: {
    options: {
      showPanel: true,
      selectedPanel: 'storybook/knobs/panel',
      previewTabs: {
        actions: {
          hidden: true,
        },
      },
    },
  },
};

export const Default = () => ({
  components: { BaseCircularGraph },
  props: {
    value: {
      type: Number,
      default: number('value', 50),
      required: true,
    },
  },
  template: `<base-circular-graph :value='this.value' />`,
});
