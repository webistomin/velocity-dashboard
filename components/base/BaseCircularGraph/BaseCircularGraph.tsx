import { VueComponent } from 'types/vue-components';
import { Component, Prop } from 'nuxt-property-decorator';
import Vue, { VNode } from 'vue';
import { nanoid } from 'nanoid';

import BaseTitle from 'components/base/BaseTitle';
import { animateValue } from '~/utils/animate-value';

import './BaseCirclularGraph.sass';

export interface IBaseCircularGraphProps {
  value: number;
  graphId?: string;
}

@Component({
  name: 'BaseCircularGraph',
})
export default class BaseCircularGraph extends VueComponent<IBaseCircularGraphProps> {
  public $refs!: Vue['$refs'] & {
    progressBar: SVGPathElement;
  };

  @Prop({ default: 0, required: true })
  private readonly value!: IBaseCircularGraphProps['value'];

  @Prop()
  private readonly graphId!: IBaseCircularGraphProps['graphId'];

  duration: number = 2000;
  percent: number = this.value / 100;
  startValue: number = 0;
  id: string = 'base-circular-graph-mask';

  public animatedValue: IBaseCircularGraphProps['value'] = 0;

  created() {
    if (this.graphId) {
      this.id = `base-circular-graph-mask-${this.graphId}`;
    } else {
      this.id = `base-circular-graph-mask-${nanoid()}`;
    }
  }

  public mounted(): void {
    animateValue.call(this, 'animatedValue', this.startValue, this.value, this.duration);
    this.startCircularProgress(this.percent, this.duration);
  }

  public startCircularProgress(percentage: number, speed: number) {
    let startTimestamp: number = 0;
    let x = this.startValue;
    const step = (timestamp: number) => {
      if (!startTimestamp) {
        startTimestamp = timestamp;
      }
      const progress = Math.min((timestamp - startTimestamp) / speed, 1);
      x = progress * percentage;
      if (progress < 1) {
        window.requestAnimationFrame(step);
        this.drawProgress(x);
      }
    };
    window.requestAnimationFrame(step);
  }

  public clamp(n: number, min: number, max: number) {
    return Math.max(min, Math.min(max, n));
  }

  public drawProgress(percent: number) {
    if (isNaN(percent)) {
      return;
    }

    percent = this.clamp(parseFloat(String(percent)), 0, 1);

    const angle = this.clamp(percent * 180, 0, 359.99999);
    const paddedRadius = 49 + 1;
    const radians = (angle * Math.PI) / 180;
    const x = Math.sin(radians) * paddedRadius;
    const y = Math.cos(radians) * -paddedRadius;
    const mid = angle > 180 ? 1 : 0;
    const pathData = 'M 0 0 v -%@ A %@ %@ 1 '.replace(/%@/gi, String(paddedRadius)) + mid + ' 1 ' + x + ' ' + y + ' z';

    const bar = this.$refs.progressBar;
    bar.setAttribute('d', pathData);
  }

  render(): VNode {
    return (
      <div class='base-circular-graph'>
        <div class='base-circular-graph__wrapper'>
          <svg
            class='base-circular-graph__progress'
            viewBox='0 0 100 100'
            width={223}
            height={223}
            shape-rendering='geometricPrecision'>
            <defs>
              <mask id={this.id} x='0' y='0' width='120' height='120' maskUnits='userSpaceOnUse'>
                <circle cx='50' cy='50' r='51' stroke-width='0' fill='black' opacity='1' />
                <circle
                  id='bar'
                  r='50'
                  cx='50'
                  cy='50'
                  fill='transparent'
                  stroke-dasharray='1'
                  stroke-dashoffset='1000'
                  stroke='white'
                  stroke-width='16'
                />
                <circle
                  class='base-circular-graph__progress-inner'
                  cx='50'
                  cy='50'
                  r='40'
                  stroke-width='0'
                  fill='black'
                  opacity='1'
                />
              </mask>
            </defs>
            <g mask={`url(#${this.id})`}>
              <circle
                class='base-circular-graph__track'
                cx='50'
                cy='50'
                r='50'
                opacity='1'
                fill='rgba(var(--color-primary), 0.2)'
              />
              <path class='base-circular-graph__bar' ref='progressBar' d='M 0 0' fill='rgb(var(--color-primary))' />
            </g>
          </svg>
        </div>
        <div class='base-circular-graph__info'>
          <BaseTitle level={2} class='base-circular-graph__title'>
            {this.animatedValue}
          </BaseTitle>
          <p class='base-circular-graph__caption caption paragraph'>
            Operating <br /> score
          </p>
        </div>
      </div>
    );
  }
}
