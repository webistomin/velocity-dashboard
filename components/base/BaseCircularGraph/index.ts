import BaseCircularGraph from './BaseCircularGraph';

export default BaseCircularGraph;
