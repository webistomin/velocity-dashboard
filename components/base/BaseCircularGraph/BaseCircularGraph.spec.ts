import { mount } from '@vue/test-utils';
import BaseCircularGraph from './BaseCircularGraph';

describe('BaseCircularGraph', () => {
  const factory = (propsData = {}, options = {}) => {
    return mount(BaseCircularGraph, {
      propsData: {
        value: 100,
        ...propsData,
      },
      ...options,
    });
  };

  it('Matches snapshot', () => {
    const wrapper = factory({
      graphId: 'id',
    }).html();
    expect(wrapper).toMatchSnapshot();
  });

  it('Set random id if graphId prop is not provided', () => {
    const wrapper = factory();
    const initialId = 'base-circular-graph-mask';
    const id = wrapper.vm.id;
    expect(id).not.toEqual(initialId);
  });

  it('Clamp value correctly', () => {
    const wrapper = factory();
    const result = wrapper.vm.clamp(0.3423423, 0, 1);
    expect(result).toEqual(0.3423423);
  });

  it('Draw progress correctly', () => {
    const expectedPath = 'M 0 0 v -50 A 50 50 1 0 1 50 -3.061616997868383e-15 z';
    const wrapper = factory();

    wrapper.vm.drawProgress(0.5);
    const bar = wrapper.vm.$refs.progressBar as SVGPathElement;
    const dAttr = bar.getAttribute('d');
    expect(dAttr).toEqual(expectedPath);
  });

  it('Return undefined and does not change path if number is not correct', () => {
    const wrapper = factory();
    const initialPath = 'M 0 0';

    const result = wrapper.vm.drawProgress(('abc' as unknown) as number);
    const bar = wrapper.vm.$refs.progressBar as SVGPathElement;
    const dAttr = bar.getAttribute('d');

    expect(dAttr).toEqual(initialPath);
    expect(result).toEqual(undefined);
  });
});
