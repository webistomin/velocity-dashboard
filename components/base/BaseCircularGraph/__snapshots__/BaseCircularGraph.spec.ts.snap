// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseCircularGraph Matches snapshot 1`] = `
"<div class=\\"base-circular-graph\\">
  <div class=\\"base-circular-graph__wrapper\\"><svg viewBox=\\"0 0 100 100\\" width=\\"223\\" height=\\"223\\" shape-rendering=\\"geometricPrecision\\" class=\\"base-circular-graph__progress\\">
      <defs>
        <mask id=\\"base-circular-graph-mask-id\\" x=\\"0\\" y=\\"0\\" width=\\"120\\" height=\\"120\\" maskUnits=\\"userSpaceOnUse\\">
          <circle cx=\\"50\\" cy=\\"50\\" r=\\"51\\" stroke-width=\\"0\\" fill=\\"black\\" opacity=\\"1\\"></circle>
          <circle id=\\"bar\\" r=\\"50\\" cx=\\"50\\" cy=\\"50\\" fill=\\"transparent\\" stroke-dasharray=\\"1\\" stroke-dashoffset=\\"1000\\" stroke=\\"white\\" stroke-width=\\"16\\"></circle>
          <circle cx=\\"50\\" cy=\\"50\\" r=\\"40\\" stroke-width=\\"0\\" fill=\\"black\\" opacity=\\"1\\" class=\\"base-circular-graph__progress-inner\\"></circle>
        </mask>
      </defs>
      <g mask=\\"url(#base-circular-graph-mask-id)\\">
        <circle cx=\\"50\\" cy=\\"50\\" r=\\"50\\" opacity=\\"1\\" fill=\\"rgba(var(--color-primary), 0.2)\\" class=\\"base-circular-graph__track\\"></circle>
        <path d=\\"M 0 0\\" fill=\\"rgb(var(--color-primary))\\" class=\\"base-circular-graph__bar\\"></path>
      </g>
    </svg></div>
  <div class=\\"base-circular-graph__info\\">
    <h2 class=\\"title base-title base-title_level-2  base-circular-graph__title\\">0</h2>
    <p class=\\"base-circular-graph__caption caption paragraph\\">Operating <br> score</p>
  </div>
</div>"
`;
